src/sagemaker/tensorflow/deploying_tensorflow_serving.rst

```diff
@@ -269,6 +269,194 @@ More information on how to create ``export_outputs`` can be found in `specifying
+   service, for example,
+   'http://localhost:8501/v1/models/half_plus_three:predict'
+
+-  ``grpc_uri (string)``: the GRPC port number generated by the Python
+   service, for example, '9000'
+
+-  ``custom_attributes (string)``: content of
+   'X-Amzn-SageMaker-Custom-Attributes' header from the original
+   request, for example,
+   'tfs-model-name=half*plus*\ three,tfs-method=predict'
+
+-  ``request_content_type (string)``: the original request content type,
+   defaulted to 'application/json' if not provided
+
+-  ``accept_header (string)``: the original request accept type,
+   defaulted to 'application/json' if not provided
+
+-  ``content_length (int)``: content length of the original request
+
+Here's a code example of implementing ``input_handler`` and
+``output_handler``. By providing these, the Python service will post the
+request to TFS REST uri with the data pre-processed by ``input_handler``
+and pass the response to ``output_handler`` for post-processing.
+
+.. code::
+
+   import json
+
+   def input_handler(data, context):
+       """ Pre-process request input before it is sent to TensorFlow Serving REST API
+       Args:
+           data (obj): the request data, in format of dict or string
+           context (Context): an object containing request and configuration details
+       Returns:
+           (dict): a JSON-serializable dict that contains request body and headers
+       """
+       if context.request_content_type == 'application/json':
+           # pass through json (assumes it's correctly formed)
+           d = data.read().decode('utf-8')
+           return d if len(d) else ''
+
+       if context.request_content_type == 'text/csv':
+           # very simple csv handler
+           return json.dumps({
+               'instances': [float(x) for x in data.read().decode('utf-8').split(',')]
+           })
+
+       raise ValueError('{{"error": "unsupported content type {}"}}'.format(
+           context.request_content_type or "unknown"))
+
+
+   def output_handler(data, context):
+       """Post-process TensorFlow Serving output before it is returned to the client.
+       Args:
+           data (obj): the TensorFlow serving response
+           context (Context): an object containing request and configuration details
+       Returns:
+           (bytes, string): data to return to client, response content type
+       """
+       if data.status_code != 200:
+           raise ValueError(data.content.decode('utf-8'))
+
+       response_content_type = context.accept_header
+       prediction = data.content
+       return prediction, response_content_type
+
+There are occasions when you might want to have complete controls over
+request. For example, making TFS request (REST or GRPC) to first model,
+inspecting results and making request to a second model. In this case,
+you may implement the ``handler`` instead of the ``input_handler`` and
+``output_handler`` pair:
+
```

Comment:
You might want to have complete control over the request. For example, you might want to make a TFS request (REST or GRPC) to the first model, inspect the results, and then make a request to a second model. In this case, implement the ``handler`` method instead of the ``input_handler`` and ``output_handler`` methods, as demonstrated in the following code: